source/common/stats/symbol_table_impl.h

```diff
@@ -0,0 +1,99 @@
+#pragma once
+
+#include <algorithm>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "envoy/common/exception.h"
+#include "envoy/stats/symbol_table.h"
+
+#include "common/common/assert.h"
+
+#include "absl/strings/str_join.h"
+#include "absl/strings/str_split.h"
+
+namespace Envoy {
+namespace Stats {
+
+class SymbolTableImpl : public SymbolTable {
+public:
+  std::vector<Symbol> encode(const std::string& name) override {
+    std::vector<Symbol> symbol_vec;
+    std::vector<std::string> name_vec = absl::StrSplit(name, '.');
+    symbol_vec.resize(name_vec.size());
+    std::transform(name_vec.begin(), name_vec.end(), symbol_vec.begin(),
+                   [this](std::string x) { return this->toSymbol(x); });
+    return symbol_vec;
+  }
+
+  std::string decode(const std::vector<Symbol>& symbol_vec) const override {
+    std::vector<std::string> name;
+    name.resize(symbol_vec.size());
+    std::transform(symbol_vec.begin(), symbol_vec.end(), name.begin(),
+                   [this](Symbol x) { return this->fromSymbol(x); });
+    return absl::StrJoin(name, ".");
+  }
+
+  bool free(const std::vector<Symbol>& symbol_vec) override {
+    bool successful_free = true;
+    for (const Symbol symbol : symbol_vec) {
+      successful_free = successful_free && freeSymbol(symbol);
+    }
+    return successful_free;
```

Comment:
I think we should architect this to avoid needing a free().  I thought this should all be automatic by reference-counting.

Reply:
I would prefer this approach as well.  You can build an object to wrap the vector that does the free in its destructor.

Reply:
See my above point on RAII and `SharedPtr`.